### ISSUE TITLE
Assign ECS cpu/memory in tags on a per-asset basis

### ISSUE THREAD

**Issue by skittlejohn** [dagster-io/dagster#20899]
### What's the use case?

With the open source Dagster deployed to ECS with the ECSRunLauncher, memory and compute resources can be explicitly specified in one of two ways: in the `dagster.yaml` file or as job tags. Tagging individual assets with requisite ECS resources is supported neither through the `op_tags` kwarg nor the new `tags` kwarg. This means that to configure those resources, you have to set _all_ runs to use the same resources or limit yourself to only materializing assets via jobs, effectively taking auto-materializations and manual materializations through the web UI off of the table.

### Ideas of implementation

Ideally those resources could be optionally set on each asset, with the job tags taking precedence. If we have an asset group like 
```
@asset(tags={"ecs/cpu": "1024", "ecs/memory": "8192"})
def thing_one():
    ...

@asset(tags={"ecs/cpu": "2048", "ecs/memory": "16384"})
def thing_two(thing_one):
    ...

thing_job = define_asset_job(tags={"ecs/cpu": "4096", "ecs/memory": "16384", ...)
```
the resource tag hierarchy would be defaults from `dagster.yaml`, then the individual tags, then the job tags. If no job tags are specified, the ECSRunLauncher should take the largest resource set from the individual assets and use those for the run.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by leejlFG** [comment]
I would happily work on this with a nudge in the right direction.

**Comment by yuhan** [comment]
Would this work for you? https://github.com/dagster-io/dagster/discussions/14544

**Comment by ammerzon** [comment]
This is a significant drawback for us because we would have to apply the highest hardware requirements of all assets to every run, regardless of whether they're needed.

What I don't understand is why this isn't working as expected. Shouldn't the tags be applied with the [`get_cpu_and_memory_overrides`](https://github.com/dagster-io/dagster/blob/456494d07413ebc781c08bcdfb22b69e419410d1/python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py#L398) function?

![CleanShot 2024-05-27 at 17 05 19@2x](https://github.com/dagster-io/dagster/assets/7319168/1a01c366-519d-4ae6-b6f9-f244198d635c)

**Comment by ammerzon** [comment]
@leejlFG, I see your point about these being only `run_tags`.

I hadn't considered that aspect before, good point 🤔 An `ECSStepLauncher` would be nice to have, but it's not a requirement for me. It's acceptable if all assets for the same run share the same resources.

What I care most about is having some sort of tag per asset so that auto-materializations/ad-hoc runs are launched with the necessary resources.

**Comment by jessebrowning** [comment]
Does not seem to be working when using `op_tags` with `dlt_assets`. Is this still expected to function?

**Comment by skittlejohn** [comment]
Looks like this may be possible now by adding an explicit `AutomationConditionSensorDefinition` to group assets that need the same resources. I'll test it all out next week, because this may have unlocked declarative automation in ECS for me. [AutomationConditionSensorDefinition](https://docs.dagster.io/guides/automate/declarative-automation/automation-condition-sensors#adding-run-tags)

**Comment by aksestok** [comment]
Not exactly @skittlejohn's ask, but using [the ECS executor](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py) solves the "set individual task reources per asset / op" at least. Released a year ago in [1.9.7](https://github.com/dagster-io/dagster/releases/tag/1.9.7).

```python
import dagster as dg
from dagster_aws.ecs import ecs_executor


@dg.asset(
    op_tags={  # Override resources for an asset
        "ecs/cpu": 2048,
        "ecs/memory": 16384,
        "ecs/ephemeral_storage": 100,
    }
)
def my_asset(): ...


executor = ecs_executor.configured(
    {"cpu": 1024, "memory": 2048}  # Set default resources for an op / asset
)

defs = dg.Definitions(assets=[my_asset], executor=executor)
```

**Comment by skittlejohn** [comment]
Yes, although the `ecs_executor` in combination with `AutomationConditionSensorDefinition` groupings technically cover my use cases, they remain cumbersome to implement in combination with assets and jobs that benefit from the default `multiprocess_executor`.

This primarily comes from the fact that the `AutomationConditionSensorDefinition` does not accept an executor parameter to apply to the job created behind the scenes. Therefore asset-jobs defined with `define_asset_job` can benefit from the ad hoc application of the `ecs_executor` and per-asset resource tags, but for assets run with ephemeral jobs built from the `AutomationConditionSensorDefinition`, the entire code location must be set to use the `ecs_executor` as the default executor. That would mean adding `executor_def=multiprocess_executor` to every existing job that does _not_ benefit from the `ecs_executor`.

Is that a total dealbreaker? No, but ensuring all of our users know the difference and can apply the `multiprocess_executor` on their own? Harder than the refactor itself by a long shot.

Again, technically what I want is possible, just not convenient. My ideal solution would be adding the `executor_def` arg to the `AutomationConditionSensorDefinition` so the jobs they create use it. Even better would be to set different defaults for different types of materialization strategies - all asset-jobs use the `multiprocess_executor` and all materializations resulting from the  `AutomationConditionSensorDefinition` jobs use the `ecs_executor`.

And _yes_, I could do this myself by applying a function to every asset-job before they are given to the `Definitions` object - take the existing job (which would be using the `ecs_executor` in this case to make `AutomationConditionSensorDefinition` "jobs" work correcetly) and replacing it with a new job where the `multiprocess_executor` is applied...

This is some real nitpicky stuff, ya'll can close this issue. I would appreciate an `executor_def` arg on the `AutomationConditionSensorDefinition` though.

**Comment by jonathanunderwood** [comment]
@skittlejohn has perfectly captured a real need i.e. the ability to set the executor for a declarative automation. Without this, it's not really feasible to move some work away from jobs and onto declarative automations, despite declarative automations being the preferred orchestration tool moving forwards.

**Comment by cmhac** [comment]
Bumping — this would be extremely useful for our our purposes. Currently not being able to set per-asset resources is a real limitation
